### ISSUE TITLE
RCP firmware settings.

### ISSUE THREAD

**Issue by MattWestb** [skgsergio/silabs-multiprotocol-firmware-zbgw04-usb#1]
If you is reading all post in the xsp1989 github issues you is finding that the user xsp1989 have moving on to one other company after having supporting the community with his factory cooked firmware. Hi was the only person that was having the signing key for signing the Sonoff ZBB firmware and i think we is not getting any more of them from Sonoff in the future.

Hi was also saying that the zbgw04 is having PCB traces for hardware flow control but i was not asking hi witch pins is using on both sides that is needed to know for getting it working OK.
All other Sonoff products dot have it so its only for your zbgw.

Then Silabs have moving to use all code compatible with Open thread then have ditching software flow control = bad for hardware that is not having  hardware flow control.

801.15.4 is having max 256 Kbaud in the radio / network layer and with overhead in the secondary serial protocol is better using higher baud rate but is normally needing flow control for working OK. 
Until now i was working ruing at 230400 with  HW but the 4.1.4 is getting RCR error on most my hardware and i think its one software defect or locking in the firmware that is ding it, so running all my RCP with 115200 and HW then its possible.

I think its best doing normal 115200 with no flow control so all users can using the firmware and if you like testing one with HW and higher baud rate for your zbgw.

Thanks for contributing to our IOT community !!

Mattias

Edit: I was posting one new issue in xsp1989 github and asking for the hardware settings and hoping getting one replay (if not asking you cant getting any yes) but i can being hi is not having the information now.

**Comment by skgsergio** [comment]
I think he said that a new revision of the ZB-GW04 has/will have traces for hardware control flow.

I've checked in the one I have:
 
![image](https://user-images.githubusercontent.com/186349/214864513-4fe0d913-1904-419f-b3d2-db276f484151.png)

![image](https://user-images.githubusercontent.com/186349/214864934-668a1451-5bf5-4464-acde-16ec92362001.png)

And there are no connected taces to RTS/CTS. I've also checked for continuity from those two pins to all the pins of the ZYZBP008 SM-011 module and they are not connected.

**Comment by MattWestb** [comment]
Sad :-(((
Perhaps it was the last version of the ZBGW  that is using unsigned firmware (i think its branded ewlink) so not much to do for the moment.
If someone is getting with pad on the module is using for it we can documenting it and then is also doing little hardware mod on your device for getting the right pads connected from the CH340.

PS The CH340 chip can doing 150000 baud without flow control and i is using it in may devices.

**Comment by MattWestb** [comment]
I was getting one answer from from XPS1989 and PD03-RTS and PD04-CTS if you like cooking one with s_tandard_ HW flow control and you can also doing the connections on your dongle and you is getting better / more stable continuation with the chip.
I have soldering SWD cables on my preserie SkyConnect and its was very small pads and not easy to do and i think its the same on you dongle.
 
https://github.com/xsp1989/zigbeeFirmware/issues/34#issuecomment-1406527600

**Comment by xsp1989** [comment]
We have updated the connection of CTS/RTS in the latest version of hardware, the latest version of hardware is v1.2, and other older versions do not support hardware connection.

**Comment by skgsergio** [comment]
@xsp1989 thanks for the info, I'll try to get a 1.2 revision, but in a quick AliExpress search I only found 1.1 ones.

@MattWestb I thought about doing a hardware mod but my soldiering skills are not enough for that kind of spacing between the pins of the CH340, also I'm not sure if they are grounded which would mean it requires unsoldering from the PCB.

I might build firmwares with HW Flow Control and without it and adding the relevant info about which version should you use depending on your device revision.

**Comment by xsp1989** [comment]
@skgsergio https://www.aliexpress.us/item/3256802605351277.html
This link must be V1.2, and there are still some sellers selling it, which is also v1.2, but the picture may not be updated, you can communicate with the seller

**Comment by skgsergio** [comment]
@xsp1989 ordered from there, they confirmed me via chat it is V1.2

I'll link that store in my README when it arrives if everything is ok.

Thanks!

@MattWestb I uploaded yesterday (untested) builds for 1.2, but they should work as the only difference is enabling hardware control flow with the pins PD03 for RTS and PD04 for CTS while keeping the baudrate as 115200. Once I have my new V1.2 I'll test them and maybe check if it is possible to raise the baudrate to 230400 with flow control.

**Comment by skgsergio** [comment]
V1.2 is here, was pretty quick.

I've found that although the PCB says PD04/CTS and PD03/RTS you need to build the firmware setting CTS as PD03 and RTS as PD04.

I've tested it a bit but I won't migrate my current setup to the new device until tomorrow at least (I don't want to break light-related stuff at night 😹).

**Comment by MattWestb** [comment]
It can being they have writing the signal that shall being connected (as input) and not the out put like TX > RX and RX > TX. If its working it shall being right.
The CH340 shall not having any problems with higher baud rate and you need at least 230K if getting many device that spamming the mesh and thread devices is doing there things.

If you can always do some light bound with one remote (per device or group depends of the remote) so you can getting little light if the system is down and you can still controlling them (its longer dark in the north).
(Edit its binding the remote to device or group not the opposite !!)

Its looks like addon for 4.2.1 is being released very soon and i was having problem with the CPC communication 4.1.4 and was working OK on the 4.2.0 so i like getting it **FAST !!**

**Comment by skgsergio** [comment]
I'll experiment with higher baud rates for the hardware flow version once I have everything working.

I only have one IKEA remote but without a battery :(

**Comment by MattWestb** [comment]
You is doing it right one step at the time and see if its working OK !!!!

If the IKEA remote is on old firmware you can adding your light in one ZHA group and bond the remote to it and its working also the the system is DEAD.
If running the 2 latest firmware you must binding / device (i think max devices) and its also working OK.

If its not **too** cold in Mad you can getting one battery in the supermarcado for saving you night in the dark  ;-))

**Comment by MattWestb** [comment]
SL addon 0.13.0 up and running with 2* arm32 and 1 * arm64 and RCP 4.2.1.0.
arm64 is having 30 Zigbee device and all looks working well.
Was using 115K HW for the 4.1.4 for not getting CRC problems but 4.2.1.0 looks working good with 230K HW and i shall trying 460K HW and see if its working OK

**Comment by skgsergio** [comment]
@MattWestb your CRC problems were showing in the SL addon logs or in the Home Assistant Logs? So I can keep an eye on it when I start doing baud rate tests this evening.

I've already migrated all my setup to the ZB-GW04 v1.2 using SL addon 0.13.0 and 4.2.1 firmware with hardware flow and 115k. So far looks good, no surprises as it was working at 115k without hardware flow!

**Comment by MattWestb** [comment]
So i was to tired yesterday and was missing the upgrade and is now the last doing it ;-(

In the addon log you is getting errors and warnings and in the end of the line its written **CRC** so its easy to see.

With 4.1.1.0 it was little tricky then OTBR was working OK and EZSP receiving commands but was having problems sending commands from ZHA (I think it was not real CRC errors it was somthing wrong in the CPC communication that was doing it not working OK).

If the chips on both sides is good made and the cable is OK it shall not being any problems and Flow control only if buffer start being full its pausing the filling it up until its being lowed. If the line between is not good flow control is not helping and you is getting CRC errors.

460K HW is not working well im my setups (CRC errors) and i dont have so good quality on the CH340 and USB is not the best but im very happy with 230K !!

**Comment by skgsergio** [comment]
Looks like 460800 baud rate is giving me checksum errors, I'll stick with 230400:

```
[22:27:04:321279] WARNING : In function 'core_process_rx_s_frame' in file /usr/src/cpc-daemon/server_core/core/core.c at line #790 : Remote received a packet with an invalid checksum
[22:27:04:348998] WARNING : In function 'core_process_rx_s_frame' in file /usr/src/cpc-daemon/server_core/core/core.c at line #790 : Remote received a packet with an invalid checksum
[22:27:04:376614] WARNING : In function 'core_process_rx_s_frame' in file /usr/src/cpc-daemon/server_core/core/core.c at line #790 : Remote received a packet with an invalid checksum
[22:27:04:417689] WARNING : In function 'core_process_rx_s_frame' in file /usr/src/cpc-daemon/server_core/core/core.c at line #790 : Remote received a packet with an invalid checksum
```

**Comment by skgsergio** [comment]
It's been a couple of days with the 230400 baud rate and hardware flow without issues, I'm going to close this issue and keep building 115200 for the non-hardware flow version and 230400 for the hardware flow version.

**Comment by MattWestb** [comment]
230K shall being enough for normal Zigbee and thread and if its stable use it for your systems.
For the no Flow control 115K shall being OK and if the user like having higher speed then upgrade hardware.

Great work done !!

I posting if i finding somthing interesting  for you :-)))

**Comment by MattWestb** [comment]
By the was i have linking you git for Sonoff RCP firmware from https://github.com/MattWestb/EFR32-FW then i finding it very well don with instructions and all !!

**Comment by MattWestb** [comment]
I have some nice new Zifbee modules with EFR32MG21A020F1024IM32 with "factory" EZSP 6.7.8 and its working OK.

Was trying cooking RCP and EZSP for it and i getting RCP working well with and can communicating with universal-silabs-flasher in bootloader mode and RCP with hardware and without flow control = OK.

Was trying cocking EZSP 7.2.1.0 and 6.10.3.0 with SW flow control (the last looks being broken after last SS5 update for MG1X devices) and converting s37 to GBL and flashing it with USF but i cant communicating with it. First i was thinking the communication pins was wrong but then getting RCP working it must being OK.

Do you knowing if its somthing more that i must  doing for getting EZSP it working ?
I have tuya MG21 module working with 4.X RCP and 7.X EZSP but is made with older GSDK / SS5.

Thanks in advance for hints !!

**Comment by skgsergio** [comment]
I've built `ncp-uart-hw` in SS5 with software flow control from Gecko SDK 4.2.1 which I think is EZSP 7.2.1.0 for my ZB-GW04 v1.1 (I'm using it for sniffing Zigbee using `bellows dump`) and I flashed it using the `universal-silabs-flasher` over the RCP firmware I had on it without problems...

If RCP is working the TX/RX pins should be ok yes.

The only thing I can think of is that you should ensure you have the `--baudrate` correctly set to the baud rate the device currently has (I guess you already know this as you are no stranger to these devices) and if it doesn't work then trying to reset it into bootloader and flash it setting `--probe-method bootloader` (make sure you have `0.0.9` version).

**Comment by MattWestb** [comment]
Its somthing very spooky and i was having it before and its likely corrupted files that is making it so i deleting SS5 and reinstalling it but its taking some time.
Its famous having problems with SS under windows with the emulated Linux system under that have problems with the file system.

Thanks for the feedback and i reporting back if its working OK or not.

**Comment by MattWestb** [comment]
Now i have getting it working !!
First uninstall and new install SS5 was OK but guthub mirror `FRA` (I think Frankfurt) was badly up and was taking over one hour then open one new project and adding hardware then it was not getting the hardware properties from github.

RCP 4.2.1 is working OK.
NCP 7.2.1.0 is working OK
NCP 6.10.3.0 is working OK.

But **if** installing one NCP 7.0.X and then 6.10.3.0 the 6.10.3.0 is not starting and must doing reset and force bootloader. Its one known problem then NVM3 is not OK after 7.0.X. Doing one chip erase and installing one bootloader and then from it 6.10.3.0 and its working.
Dumping the main flash minus bootloader and converting it to GBL (1 Meg large) and i can recover from bootloader.

Now i can making more tests and tuning the NCP :-))))

**Comment by MattWestb** [comment]
Silabs have released GSDK 4.2.2.0 and have fixed one bug that can helping your RCP firmware then you is tuning the HXO and they was miss tuning the UART speed > problems with comports stability.

Its described in the release notes but i did not remember if it was in GSDK, EZSP or some other one.

The problem is you cant use the RCP 4.2.2.0 until multi protocol is updated to same version.

I hope your EZSP and RCP s working god for you and other user that is using it !!!

I have getting 2 EVE energy and one Nanoleaf light joining my thread network with Home Kit Client in HA and they is very stable and much faster and reachable as the was with BT connection. Only problem i can only have one RCP as boarder router then getting routing problems if more is online.



Edit: Was finding it !!
```
1026022 Fixed an issue that was affecting the UART baud rate when setting the CTUNE value on NCP from host using set the
EZSP_CONFIG_CTUNE_VALUE command.
```
